### PR TITLE
PR: Save a reference to messagebox in EditorStack to avoid memory to be freed.

### PR DESCRIPTION
**Pull request by rlaverde**
Fixes: #4453 
Fixes: #4442 

I tried different solutions, like returning the focus, or checking if the widget is alive, but didn't work

This solution works better, and I think that's a cleaner solution (thanks @goanpeca for the idea)

Only two of the message box was related two the errors mentioned above, but I also change the others because they could raise the same error (but that use cases are less common)

### CODE REVIEW

**Comment by ccordoba12** [comment]
There's a syntax error in Python 2, that's why our tests are failing.

File: spyder/widgets/editor.py [hunk 1]
```diff
@@ -1424,11 +1427,17 @@ def save_if_changed(self, cancelable=False, index=None):
...
+                self.msgbox = QMessageBox(
+                        QMessageBox.Question,
+                        self.title,
+                        _("<b>%s</b> has been modified."
+                          "<br>Do you want to save changes?"
+                         ) % osp.basename(finfo.filename),
+                          buttons,
+                          parent=self)
+
+                answer = self.msgbox.exec()
```
Comment (goanpeca) [inline spyder/widgets/editor.py]: @rlaverde in PyQt, we always use `exec_()` with the `underscore`

File: spyder/widgets/editor.py [hunk 2]
```diff
@@ -1488,11 +1497,15 @@ def save(self, index=None, force=False):
...
-                                        str(error)))
+            self.msgbox = QMessageBox(
+                    QMessageBox.Critical,
+                    _("Save"),
+                    _("<b>Unable to save file '%s'</b>"
+                      "<br><br>Error message:<br>%s"
+                      ) % (osp.basename(finfo.filename),
+                                        str(error)),
+                      self)
+            self.msgbox.exec()
```
Comment (goanpeca) [inline spyder/widgets/editor.py]: @rlaverde in PyQt, we always use `exec_()` with the `underscore`

File: spyder/widgets/editor.py [hunk 3]
```diff
@@ -1577,11 +1590,15 @@ def save_copy_as(self, index=None):
...
-                                            str(error)))
+                self.msgbox = QMessageBox(
+                    QMessageBox.Critical,
+                    _("Save"),
+                    _("<b>Unable to save file '%s'</b>"
+                      "<br><br>Error message:<br>%s"
+                      ) % (osp.basename(finfo.filename),
+                                        str(error)),
+                      self)
+                self.msgbox.exec()
```
Comment (goanpeca) [inline spyder/widgets/editor.py]: @rlaverde in PyQt, we always use `exec_()` with the `underscore`

File: spyder/widgets/editor.py [hunk 4]
```diff
@@ -1750,12 +1767,16 @@ def __check_file_status(self, index):
...
+            self.msgbox = QMessageBox(
+                    QMessageBox.Warning,
+                    self.title,
+                    _("<b>%s</b> is unavailable "
+                      "(this file may have been removed, moved "
+                      "or renamed outside Spyder)."
+                      "<br>Do you want to close it?") % name,
+                    QMessageBox.Yes | QMessageBox.No,
+                    self)
+            answer = self.msgbox.exec()
```
Comment (goanpeca) [inline spyder/widgets/editor.py]: @rlaverde in PyQt, we always use `exec_()` with the `underscore`

File: spyder/widgets/editor.py [hunk 5]
```diff
@@ -1769,12 +1790,15 @@ def __check_file_status(self, index):
...
-                                QMessageBox.Yes | QMessageBox.No)
+                    self.msgbox = QMessageBox(
+                        QMessageBox.Question,
+                        self.title,
+                        _("<b>%s</b> has been modified outside Spyder."
+                          "<br>Do you want to reload it and lose all "
+                          "your changes?") % name,
+                        QMessageBox.Yes | QMessageBox.No,
+                        self)
+                    answer = self.msgbox.exec()
```
Comment (goanpeca) [inline spyder/widgets/editor.py]: @rlaverde in PyQt, we always use `exec_()` with the `underscore`

File: spyder/widgets/editor.py [hunk 6]
```diff
@@ -1876,11 +1900,15 @@ def revert(self):
...
-                                  QMessageBox.Yes|QMessageBox.No)
+            self.msgbox = QMessageBox(
+                    QMessageBox.Warning,
+                    self.title,
+                    _("All changes to <b>%s</b> will be lost."
+                      "<br>Do you want to revert file from disk?"
+                      ) % osp.basename(filename),
+                    QMessageBox.Yes | QMessageBox.No,
+                    self)
+            answer = self.msgbox.exec()
```
Comment (goanpeca) [inline spyder/widgets/editor.py]: @rlaverde in PyQt, we always use `exec_()` with the `underscore`

File: spyder/widgets/editor.py [hunk 7]
```diff
@@ -2032,11 +2060,15 @@ def load(self, filename, set_current=True):
...
-                                QMessageBox.Ok)
+            self.msgbox = QMessageBox(
+                    QMessageBox.Warning,
+                    self.title,
+                    _("<b>%s</b> contains mixed end-of-line "
+                      "characters.<br>Spyder will fix this "
+                      "automatically.") % name,
+                    QMessageBox.Ok,
+                    self)
+            self.msgbox.exec()
```
Comment (goanpeca) [inline spyder/widgets/editor.py]: @rlaverde in PyQt, we always use `exec_()` with the `underscore`

File: spyder/widgets/editor.py (hunk 1 above)
Comment (rlaverde) [inline spyder/widgets/editor.py]: Thanks, sometimes I forget about python2.7 :(

File: spyder/app/tests/test_mainwindow.py [hunk 8]
```diff
@@ -672,6 +672,29 @@ def test_open_files_in_new_editor_window(main_window, qtbot):
 
 
 @flaky(max_runs=3)
+def test_close_when_file_is_changed(main_window, qtbot):
```
Comment (ccordoba12) [inline spyder/app/tests/test_mainwindow.py]: This test is hanging in Travis with the pip wheels. So please skip it when `PYQT_WHEEL`

**Comment by goanpeca** [comment]
@ccordoba12 merging?

**Comment by ccordoba12** [comment]
Yep. Thanks @rlaverde, great work!

**Pull request merged**